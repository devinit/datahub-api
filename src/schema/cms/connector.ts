import axios from 'axios';
import * as R from 'ramda';
import * as LRU from 'lru-cache';
import * as converter from 'csvtojson';
import {queue} from '../../lib/cache';

const baseUrl = 'https://raw.githubusercontent.com/devinit/datahub-cms/master';

const createUrl: (endPoint: string) => string = (endPoint) => `${baseUrl}/${endPoint}`;

const httpGet: (api: string) => Promise < string > = R.composeP(R.prop('data'), axios.get);

const lruOpts: LRU.Options<any> = {
    max: 300,
    maxAge: 1000 * 60 * 60 * 24 * 60 // TODO: create time constant (60 days -- 2 months)
};

export const cache: LRU.Cache<any> = LRU(lruOpts);

export const csvToJson = <T extends {}> (csvStr: string): Promise<T[]>  =>
    new Promise(<T>(resolve, reject) => {
        const data: T[] = [];
        converter({workerNum: 2})
        .fromString(csvStr)
        .on('json', (json) => {
            data.push(json);
        })
        .on('done', (error) => {
            resolve(data);
            reject(error);
        });
    });

export const get = async <T extends {}> (endPoint: string): Promise <T[]> => {
    const api = createUrl(endPoint);
    if (cache.has(endPoint))  {
        // add to queue so that we always have freshest data
        queue(endPoint, 'cms', cache, get); // makes same query in 15 minutes so as to update cache
        return cache.get(endPoint) as T[];
    }
    const csvStr = await httpGet(api); // TODO: if github is down, fetch from a cache dumb
    const data: T[] = await csvToJson<T>(csvStr);
    cache.set(endPoint, data);
    return data;
};
