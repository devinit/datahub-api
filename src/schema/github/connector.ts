import axios from 'axios';
import * as R from 'ramda';
import * as LRU from 'lru-cache';
import * as converter from 'csvtojson';

const baseUrl = 'https://raw.githubusercontent.com/devinit/datahub-cms/master';

const createUrl: (endPoint: string) => string = (endPoint) => `${baseUrl}/${endPoint}`;

const httpGet: (api: string) => Promise < string > = R.pipeP(axios.get, R.prop('data'));

// TODO: exported so that we can do cache state management
export const cache = LRU({
    max: 50,
    maxAge: 1000 * 60 * 60
});

export const csvToJson: <T> (csvStr: string) => Promise<T[]> = (csvStr) =>
    new Promise((resolve, reject) => {
        const data = [];
        converter({workerNum: 2})
        .fromString(csvStr)
        .on('json', (json) => {
            data.push(json);
        })
        .on('done', (error) => {
            resolve(data);
            reject(error);
        });
    });

export const get: <T> (endPoint: string) => Promise <T[]> = async (endPoint) => {
    const api = createUrl(endPoint);
    if (cache.has(endPoint)) return cache.get(endPoint);
    try {
        const csvStr = await httpGet(api);
        const data = await csvToJson(csvStr);
        cache.set(endPoint, data);
        return data;
    } catch (error) {
        console.error(error);
    }
};
