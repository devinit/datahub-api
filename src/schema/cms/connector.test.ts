import 'jest';
import {csvToJson, get} from './connector';
import * as prettyFormat from 'pretty-format';

describe('Github connector', () => {

    it('should turn csv string to json Array', async () => {
        const csvStr = `id,name
                        20,allan
                        30,alex`;
        const data = await csvToJson<{name: string; id: number}>(csvStr);
        expect(data.length).toBe(2);
        expect(data[0].name).toBe('allan');
    });

    it('should get data from github and return it as json', async () => {
        const themes = await get<{id: string}>('global-picture/theme.csv');
        expect(themes.length).toBeGreaterThan(4);
        expect(prettyFormat(themes)).toMatchSnapshot();
    }, 5000);
});
