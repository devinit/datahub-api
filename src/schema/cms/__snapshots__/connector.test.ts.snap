// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github connector should get data from github and return it as json 1`] = `
"Array [
  Object {
    \\"default\\": \\"data_series.poor_people_190\\",
    \\"id\\": \\"poverty\\",
    \\"name\\": \\"Poverty\\",
    \\"position\\": \\"1\\",
  },
  Object {
    \\"default\\": \\"data_series.fragile_states\\",
    \\"id\\": \\"vulnerability\\",
    \\"name\\": \\"Vulnerability\\",
    \\"position\\": \\"2\\",
  },
  Object {
    \\"default\\": \\"non-grant-revenue-ppp-pc\\",
    \\"id\\": \\"government-finance\\",
    \\"name\\": \\"Government finance\\",
    \\"position\\": \\"3\\",
  },
  Object {
    \\"default\\": \\"largest-intl-flow\\",
    \\"id\\": \\"international-finance\\",
    \\"name\\": \\"International finance\\",
    \\"position\\": \\"4\\",
  },
  Object {
    \\"default\\": \\"fact.in_oda_gross_2015\\",
    \\"id\\": \\"international-official-finance\\",
    \\"name\\": \\"International official finance\\",
    \\"position\\": \\"5\\",
  },
  Object {
    \\"default\\": \\"in-ha\\",
    \\"id\\": \\"humanitarian-finance\\",
    \\"name\\": \\"Humanitarian finance\\",
    \\"position\\": \\"6\\",
  },
  Object {
    \\"default\\": \\"latest-census\\",
    \\"id\\": \\"data-revolution\\",
    \\"name\\": \\"Data revolution\\",
    \\"position\\": \\"7\\",
  },
  Object {
    \\"default\\": \\"data_series.iati-budgets-gb\\",
    \\"id\\": \\"forward-looking-oda\\",
    \\"name\\": \\"Forward looking ODA\\",
    \\"position\\": \\"8\\",
  },
]"
`;
